homeassistant/components/notify/__init__.py

```diff
@@ -64,101 +66,133 @@ def send_message(hass, message, title=None, data=None):
     hass.services.call(DOMAIN, SERVICE_NOTIFY, info)
 
 
-def setup(hass, config):
+@asyncio.coroutine
+def async_setup(hass, config):
     """Setup the notify services."""
-    descriptions = load_yaml_config_file(
+    descriptions = yield from hass.loop.run_in_executor(
+        None, load_yaml_config_file,
         os.path.join(os.path.dirname(__file__), 'services.yaml'))
 
     targets = {}
 
-    def setup_notify_platform(platform, p_config=None, discovery_info=None):
+    @asyncio.coroutine
+    def async_setup_platform(p_type, p_config=None, discovery_info=None):
         """Set up a notify platform."""
         if p_config is None:
             p_config = {}
         if discovery_info is None:
             discovery_info = {}
 
-        notify_implementation = bootstrap.prepare_setup_platform(
-            hass, config, DOMAIN, platform)
+        platform = bootstrap.async_prepare_setup_platform(
+            hass, config, DOMAIN, p_type)
 
-        if notify_implementation is None:
+        if platform is None:
             _LOGGER.error("Unknown notification service specified")
-            return False
-
-        notify_service = notify_implementation.get_service(
-            hass, p_config, discovery_info)
-
-        if notify_service is None:
-            _LOGGER.error("Failed to initialize notification service %s",
-                          platform)
-            return False
-
-        def notify_message(notify_service, call):
+            return
+
+        _LOGGER.info("Setting up %s.%s", DOMAIN, p_type)
+        notify_service = None
+        try:
+            if hasattr(platform, 'async_get_service'):
+                notify_service = yield from \
+                    platform.async_get_service(
+                        hass, p_config, discovery_info)
+        elif hasattr(platform, 'get_service'):
+                notify_service = yield from hass.loop.run_in_executor(
+                    None, platform.get_service, hass, p_config,
+                    discovery_info)
+            else:
+                raise HomeAssistantError("Invalid notify platform.")
+
+            if notify_service is None:
+                _LOGGER.error(
+                    "Failed to initialize notification service %s", p_type)
+                return
+
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception('Error setting up platform %s', p_type)
+            return
+
+        notify_service.hass = hass
```

Comment:
Yes, it's for the `BaseNotificationService`, for times when the `async_send_message` method won't be overriden, right? But `hass` is usually passed to the NotificationService instance in the platform at init, so this is sort of doing that twice. But I guess it's the simplest way of making sure `hass` is always available in the `BaseNotificationService`.